stdlib/LinearAlgebra/src/diagonal.jl

```diff
@@ -659,10 +659,20 @@ end
 # disambiguation methods: * of Diagonal and Adj/Trans AbsVec
 *(x::Adjoint{<:Any,<:AbstractVector}, D::Diagonal) = Adjoint(map((t,s) -> t'*s, D.diag, parent(x)))
 *(x::Transpose{<:Any,<:AbstractVector}, D::Diagonal) = Transpose(map((t,s) -> transpose(t)*s, D.diag, parent(x)))
-*(x::Adjoint{<:Any,<:AbstractVector}, D::Diagonal, y::AbstractVector) =
-    mapreduce(t -> t[1]*t[2]*t[3], +, zip(x, D.diag, y))
-*(x::Transpose{<:Any,<:AbstractVector}, D::Diagonal, y::AbstractVector) =
-    mapreduce(t -> t[1]*t[2]*t[3], +, zip(x, D.diag, y))
+function *(x::Adjoint{<:Any,<:AbstractVector}, D::Diagonal, y::AbstractVector)
+    if all(isempty.((x, D, y)))
+        return zero(promote_type(eltype.((x, D, y))...))
+    else
+        return mapreduce(t -> t[1]*t[2]*t[3], +, zip(x, D.diag, y))
+    end
```

Comment:
FWIW I think something like

```suggestion
    return mapreduce(t -> t[1]*t[2]*t[3], +, zip(x, D.diag, y);
                     init = zero(promote_type(map(eltype, (x, D, y))...)))
```

may be better because:

1. Specifying a concrete value for `init` helps inference with `reduce`-like code in general.
2. Using `map` instead of broadcasting is less pressure to the compiler.

Reply:
Sure, let's try this!